### PR TITLE
Adds A Progress bar

### LINKED ISSUE #20: Add a progress indicator
It can take a while to run. It would be nice to get a way to monitor progress

### PR DESCRIPTION
**Pull request by mpchadwick**
Multiple people have asked me to give them projections for how long this will take.

image

Fixes #20

### CODE REVIEW

File: Model/Runner.php [hunk 1]
```diff
@@ -35,24 +35,29 @@ public function __construct(
...
+            return $result;
+        }
+
+        foreach ($this->checkers as $type => $checker) {
+            $checked = $checker->check($pathInfo);
+            if (!empty($checked)) {
+                $result['type'] = $type;
+                $result['items'] = $checked['customized'];
+                $result['path'] = $pathInfo['fullpath'];
+                continue;
```
Comment (timothydmahon) [inline Model/Runner.php:57]: Don't think we need this `continue;`

File: Model/Runner.php [hunk 2]
```diff
@@ -35,24 +35,29 @@ public function __construct(
...
 
-    public function run($diff)
+    public function run($line)
     {
-        foreach ($diff as $line) {
-            $pathInfo = $this->lineProcessor->toPathInfo($line);
+        $result = [];
+        $result['type'] = '';
+        $result['items'] = [];
+        $result['path'] = '';
```
Comment (timothydmahon) [inline Model/Runner.php:43]: https://github.com/sdinteractive/Magento2_SomethingDigital_UpgradeHelper/blob/abfb4c1f7f5d360969a121c66df47f65e48efd51/Model/Runner.php#L18 recommend deletion of private variable as no longer used.

**Review by timothydmahon — CHANGES_REQUESTED** [review]
minor nits

File: Console/UpgradeHelperCommand.php [hunk 3]
```diff
@@ -30,7 +31,25 @@ protected function configure()
...
+        $result = [];
+        $result['preferences'] = [];
+        $result['overrides'] = [];
+
+        $progressBar = new ProgressBar($output, $lines);
+        $progressBar->setFormat('Lines in diff processed: %current%/%max% [%bar%] %percent:3s%% (elapsed: %elapsed% | remaining: ~%remaining%)');
+
+        foreach ($diff as $line) {
+            $progressBar->advance();
+            extract($this->runner->run($line));
```
Comment (timothydmahon) [inline Console/UpgradeHelperCommand.php:45]: Another thought. For readability/long term maintenance do we want to comment here on extract? 
```
/* 
* Extract will populate: 
* $type
* $path
* $items
* /
```
Essentially keep track of the intended variables extract is meant to define here.

File: Test/Unit/Model/RunnerTest.php [hunk 4]
```diff
@@ -62,7 +62,13 @@ public function testRun()
             'diff -r Magento-EE-2.3.1/vendor/magento/module-catalog-rule/Controller/Adminhtml/Promo/Catalog/Save.php Magento-EE-2.3.3/vendor/magento/module-catalog-rule/Controller/Adminhtml/Promo/Catalog/Save.php'
         ];
 
-        $result = $this->runner->run($diff);
+        $result = [];
+        $result['preferences'] = [];
+        $result['overrides'] = [];
+        foreach ($diff as $line) {
+            extract($this->runner->run($line));
```
Comment (timothydmahon) [inline Test/Unit/Model/RunnerTest.php:69]: Same comment about adding extra comments around extract to this area. 

```
/* 
* Extract will populate: 
* $type
* $path
* $items
* /
```

File: Model/Runner.php (hunk 1 above)
Comment (mpchadwick) [reply Model/Runner.php]: You're right. This should be a `break` now that the runner receives each line one at a time. The idea is that if one the checkers hits we don't need to check anymore.

File: Console/UpgradeHelperCommand.php (hunk 3 above)
Comment (mpchadwick) [reply Console/UpgradeHelperCommand.php:47]: Yeah...extract is weird and magical in a way that I don't love, but it was appropriate here.

**Review by timothydmahon — APPROVED** [review]

**Pull request merged**